ledger/participant-integration-api/src/main/scala/platform/store/backend/common/EventStorageBackendTemplate.scala

```diff
@@ -0,0 +1,464 @@
+    events(
+      columnPrefix = "",
+      joinClause = "",
+      additionalAndClause = (
+        """  event_sequential_id > {startExclusive} AND
+          |  event_sequential_id <= {endInclusiveSeq} AND""".stripMargin,
+        List(
+          "startExclusive" -> rangeParams.startExclusive,
+          "endInclusiveSeq" -> rangeParams.endInclusive,
+        ),
+      ),
+      rowParser = rawFlatEventParser,
+      selectColumns = selectColumnsForFlatTransactions,
+      witnessesColumn = "flat_event_witnesses",
+    )(
+      limit = rangeParams.limit,
+      fetchSizeHint = rangeParams.fetchSizeHint,
+      filterParams,
+    )(connection)
+  }
+
+  override def flatTransaction(
+      transactionId: TransactionId,
+      filterParams: FilterParams,
+  )(connection: Connection): Vector[EventsTable.Entry[Raw.FlatEvent]] = {
+    import com.daml.platform.store.Conversions.ledgerStringToStatement
+    events(
+      columnPrefix = "",
+      joinClause = """JOIN parameters ON
+          |  (participant_pruned_up_to_inclusive is null or event_offset > participant_pruned_up_to_inclusive)
+          |  AND event_offset <= ledger_end""".stripMargin,
+      additionalAndClause = (
+        """  transaction_id = {transactionId} AND
+          |  event_kind != 0 AND -- we do not want to fetch divulgence events""".stripMargin,
+        List(
+          "transactionId" -> transactionId
+        ),
+      ),
+      rowParser = rawFlatEventParser,
+      selectColumns = selectColumnsForFlatTransactions,
+      witnessesColumn = "flat_event_witnesses",
+    )(
+      limit = None,
+      fetchSizeHint = None,
+      filterParams = filterParams,
+    )(connection)
+  }
+
+  override def transactionTreeEvents(
+      rangeParams: RangeParams,
+      filterParams: FilterParams,
+  )(connection: Connection): Vector[EventsTable.Entry[Raw.TreeEvent]] = {
+    events(
+      columnPrefix = "",
+      joinClause = "",
+      additionalAndClause = (
+        """  event_sequential_id > {startExclusive} AND
+          |  event_sequential_id <= {endInclusiveSeq} AND
+          |  event_kind != 0 AND -- we do not want to fetch divulgence events""".stripMargin,
+        List(
+          "startExclusive" -> rangeParams.startExclusive,
+          "endInclusiveSeq" -> rangeParams.endInclusive,
+        ),
+      ),
+      rowParser = rawTreeEventParser,
+      selectColumns =
+        s"$selectColumnsForTransactionTree, ${eventStrategy.columnEqualityBoolean("event_kind", "20")} as exercise_consuming",
+      witnessesColumn = "tree_event_witnesses",
+    )(
+      limit = rangeParams.limit,
+      fetchSizeHint = rangeParams.fetchSizeHint,
+      filterParams,
+    )(connection)
+  }
+
+  // TODO the ordering of the results here is based on node_index in the previous implementation. current implementation uses event_sequential_id. Are these the same? (if not perhaps flatTransaction needs adapting too?)
```

Comment:
Did you check this TODO?

1. `event_sequential_id` is generated by the indexer and was introduced by the append-only schema, so we should be able to verify whether the ordering is the same
2. AFAICT, the ordering of events is only important when constructing the root nodes of the transaction, [see here](https://github.com/digital-asset/daml/blob/main/ledger/participant-integration-api/src/main/scala/platform/store/appendonlydao/events/EventsTable.scala#L117).
3. The JdbcLedgerDao suite has transactions with multiple root nodes, [see here](https://github.com/digital-asset/daml/blob/main/ledger/participant-integration-api/src/test/lib/scala/platform/store/dao/JdbcLedgerDaoSuite.scala#L610). Some of the tests would fail if the order of root nodes were shuffled, e.g., "filter correctly by multiple parties with the same template". The root nodes in the tests are always sorted by node id (because that's how TransactionBuilder adds them), not sure whether that is always the true and whether it affects the result.

Reply:
Wasn't the `event_sequential_id` introduced before append-only schema was a thing to ensure the correct usage of indexes for paginatation?

Reply:
@stefanobaghino-da you're right, it was introduced before as well. It slightly changed with parallel ingestion, because before it was populated by a db sequence, now scala-logic (to be able to assign these separately from actual parallel linsertion)

Reply:
So these ordering should be the same, as confirmed by @remyhaemmerle-da 


Marton Nagy Today at 3:37 PM
Question about NodeId: if we traverse a transaction with .foldInExecutionOrder will that sequence result monotonically increasing nodeIds? (eg the ordering defined by execution-order and nodeId is the same?) 

remy
Yes, if the transaction you consider is generated from the engine.
If you omit end exerciseEnd and  rollbackEnd

Marton Nagy
we do that indeed (omit) (edited)

Reply:
Removed theTODO